File: backend/app/repositories/timeline.py
from datetime import datetime, timezone

from sqlalchemy import String, and_, cast, func, or_, select
from sqlalchemy.orm import Session

from app.models import (
    HandoverShare,
    HandoverSummary,
    TimelineAiSearchAudit,
    TimelineComment,
    TimelineEntry,
    TimelineEventTypeConfig,
    TimelineRetentionAction,
    TimelineRetentionPolicy,
    TimelineTombstone,
)


class TimelineRepository:
    def __init__(self, db: Session) -> None:
        self.db = db

    def add(
        self,
        *,
        account_id: str,
        event_type: str,
        module: str,
        title: str,
        description: str,
        performed_by: str,
        performed_by_name: str,
        engagement_id: str | None = None,
        source_record_id: str | None = None,
        source_record_type: str | None = None,
        source_record_route: str | None = None,
        before_value: dict | None = None,
        after_value: dict | None = None,
        metadata: dict | None = None,
        is_sensitive: bool = False,
        sensitivity_level: str | None = None,
        event_at: datetime | None = None,
        tags: list | None = None,
        mentions: list | None = None,
        attachments: list | None = None,
        source_hash: str | None = None,
        idempotency_key: str | None = None,
        event_type_config_id: str | None = None,
        is_system_generated: bool = True,
        is_immutable: bool = True,
    ) -> TimelineEntry:
        existing = self.get_by_dedup_key(account_id=account_id, source_hash=source_hash, idempotency_key=idempotency_key)
        if existing is not None:
            return existing

        entry = TimelineEntry(
            account_id=account_id,
            engagement_id=engagement_id,
            event_type=event_type,
            module=module,
            title=title,
            description=description,
            performed_by=performed_by,
            performed_by_name=performed_by_name,
            source_record_id=source_record_id,
            source_record_type=source_record_type,
            source_record_route=source_record_route,
            before_value=before_value,
            after_value=after_value,
            metadata_json=metadata,
            is_sensitive=is_sensitive,
            sensitivity_level=sensitivity_level,
            event_at=event_at,
            tags=tags or [],
            mentions=mentions or [],
            attachments=attachments or [],
            source_hash=source_hash,
            idempotency_key=idempotency_key,
            event_type_config_id=event_type_config_id,
            is_system_generated=is_system_generated,
            is_immutable=is_immutable,
        )
        self.db.add(entry)
        self.db.flush()
        return entry

    def get_by_dedup_key(self, *, account_id: str, source_hash: str | None, idempotency_key: str | None) -> TimelineEntry | None:
        if not source_hash and not idempotency_key:
            return None
        conditions = [TimelineEntry.account_id == account_id]
        dedup_conditions = []
        if source_hash:
            dedup_conditions.append(TimelineEntry.source_hash == source_hash)
        if idempotency_key:
            dedup_conditions.append(TimelineEntry.idempotency_key == idempotency_key)
        return self.db.scalar(select(TimelineEntry).where(*conditions, or_(*dedup_conditions)))

    def get_event(self, event_id: str) -> TimelineEntry | None:
        return self.db.get(TimelineEntry, event_id)

    def get_account_event(self, account_id: str, event_id: str) -> TimelineEntry | None:
        return self.db.scalar(select(TimelineEntry).where(TimelineEntry.id == event_id, TimelineEntry.account_id == account_id))

    def list_for_account(
        self,
        account_id: str,
        *,
        search: str | None = None,
        event_types: list[str] | None = None,
        modules: list[str] | None = None,
        owner_id: str | None = None,
        source_record_type: str | None = None,
        source_record_id: str | None = None,
        status_filter: str | None = None,
        sensitivity_level: str | None = None,
        date_from: datetime | None = None,
        date_to: datetime | None = None,
        include_sensitive: bool = False,
        include_restricted: bool = False,
        sort: str = "event_at",
        direction: str = "desc",
        page: int = 1,
        page_size: int = 100,
    ) -> tuple[list[TimelineEntry], int]:
        conditions = self._timeline_conditions(
            account_id=account_id,
            search=search,
            event_types=event_types,
            modules=modules,
            owner_id=owner_id,
            source_record_type=source_record_type,
            source_record_id=source_record_id,
            status_filter=status_filter,
            sensitivity_level=sensitivity_level,
            date_from=date_from,
            date_to=date_to,
            include_sensitive=include_sensitive,
            include_restricted=include_restricted,
        )
        total = self.db.scalar(select(func.count(TimelineEntry.id)).where(*conditions)) or 0
        order_column = TimelineEntry.event_at if sort == "event_at" else TimelineEntry.created_at
        order_column = order_column.desc() if direction == "desc" else order_column.asc()
        items = list(
            self.db.scalars(
                select(TimelineEntry)
                .where(*conditions)
                .order_by(order_column, TimelineEntry.created_at.desc())
                .offset((page - 1) * page_size)
                .limit(page_size)
            )
        )
        return items, total

    def list_for_engagement(
        self,
        engagement_id: str,
        page: int = 1,
        page_size: int = 100,
        *,
        include_sensitive: bool = False,
        include_restricted: bool = False,
    ) -> tuple[list[TimelineEntry], int]:
        conditions = [TimelineEntry.engagement_id == engagement_id, TimelineEntry.status != "deleted"]
        if not include_restricted:
            conditions.append(TimelineEntry.status != "restricted")
        if not include_sensitive:
            conditions.append(TimelineEntry.is_sensitive.is_(False))
        total = self.db.scalar(select(func.count(TimelineEntry.id)).where(*conditions)) or 0
        items = list(
            self.db.scalars(
                select(TimelineEntry)
                .where(*conditions)
                .order_by(TimelineEntry.event_at.desc(), TimelineEntry.created_at.desc())
                .offset((page - 1) * page_size)
                .limit(page_size)
            )
        )
        return items, total

    def list_retention_candidates(self, policy: TimelineRetentionPolicy, *, limit: int = 500) -> list[TimelineEntry]:
        cutoff = datetime.now(timezone.utc) - policy_duration(policy)
        conditions = [
            TimelineEntry.status == "active",
            TimelineEntry.event_at <= cutoff,
        ]
        return list(
            self.db.scalars(
                select(TimelineEntry)
                .where(*conditions)
                .order_by(TimelineEntry.event_at.asc())
                .limit(limit)
            )
        )

    def save_event(self, entry: TimelineEntry) -> TimelineEntry:
        self.db.add(entry)
        self.db.flush()
        return entry

    def add_comment(self, comment: TimelineComment) -> TimelineComment:
        self.db.add(comment)
        self.db.flush()
        return comment

    def get_comment(self, comment_id: str) -> TimelineComment | None:
        return self.db.get(TimelineComment, comment_id)

    def list_comments(self, event_id: str, page: int = 1, page_size: int = 25) -> tuple[list[TimelineComment], int]:
        conditions = [TimelineComment.timeline_entry_id == event_id, TimelineComment.deleted_at.is_(None)]
        total = self.db.scalar(select(func.count(TimelineComment.id)).where(*conditions)) or 0
        items = list(
            self.db.scalars(
                select(TimelineComment)
                .where(*conditions)
                .order_by(TimelineComment.created_at.asc())
                .offset((page - 1) * page_size)
                .limit(page_size)
            )
        )
        return items, total

    def list_event_types(
        self,
        *,
        search: str | None = None,
        active_state: str = "all",
        category: str | None = None,
        module: str | None = None,
        page: int = 1,
        page_size: int = 50,
    ) -> tuple[list[TimelineEventTypeConfig], int]:
        conditions = []
        if search and search.strip():
            term = f"%{search.strip()}%"
            conditions.append(or_(TimelineEventTypeConfig.name.ilike(term), TimelineEventTypeConfig.slug.ilike(term)))
        if active_state == "active":
            conditions.append(TimelineEventTypeConfig.is_active.is_(True))
        if active_state == "inactive":
            conditions.append(TimelineEventTypeConfig.is_active.is_(False))
        if category:
            conditions.append(TimelineEventTypeConfig.category == category)
        if module:
            conditions.append(TimelineEventTypeConfig.module == module)
        total = self.db.scalar(select(func.count(TimelineEventTypeConfig.id)).where(*conditions)) or 0
        items = list(
            self.db.scalars(
                select(TimelineEventTypeConfig)
                .where(*conditions)
                .order_by(TimelineEventTypeConfig.display_order.asc(), TimelineEventTypeConfig.name.asc())
                .offset((page - 1) * page_size)
                .limit(page_size)
            )
        )
        return items, total

    def get_event_type(self, type_id: str) -> TimelineEventTypeConfig | None:
        return self.db.get(TimelineEventTypeConfig, type_id)

    def get_event_type_by_slug(self, slug: str) -> TimelineEventTypeConfig | None:
        return self.db.scalar(select(TimelineEventTypeConfig).where(TimelineEventTypeConfig.slug == slug))

    def add_event_type(self, event_type: TimelineEventTypeConfig) -> TimelineEventTypeConfig:
        self.db.add(event_type)
        self.db.flush()
        return event_type

    def list_retention_policies(
        self,
        *,
        search: str | None = None,
        active_state: str = "all",
        entity_type: str | None = None,
        page: int = 1,
        page_size: int = 25,
    ) -> tuple[list[TimelineRetentionPolicy], int]:
        conditions = []
        if search and search.strip():
            term = f"%{search.strip()}%"
            conditions.append(or_(TimelineRetentionPolicy.name.ilike(term), TimelineRetentionPolicy.reason_template.ilike(term)))
        if active_state == "active":
            conditions.append(TimelineRetentionPolicy.is_active.is_(True))
        if active_state == "inactive":
            conditions.append(TimelineRetentionPolicy.is_active.is_(False))
        if entity_type:
            conditions.append(TimelineRetentionPolicy.entity_type == entity_type)
        total = self.db.scalar(select(func.count(TimelineRetentionPolicy.id)).where(*conditions)) or 0
        items = list(
            self.db.scalars(
                select(TimelineRetentionPolicy)
                .where(*conditions)
                .order_by(TimelineRetentionPolicy.created_at.desc())
                .offset((page - 1) * page_size)
                .limit(page_size)
            )
        )
        return items, total

    def get_retention_policy(self, policy_id: str) -> TimelineRetentionPolicy | None:
        return self.db.get(TimelineRetentionPolicy, policy_id)

    def add_retention_policy(self, policy: TimelineRetentionPolicy) -> TimelineRetentionPolicy:
        self.db.add(policy)
        self.db.flush()
        return policy

    def add_retention_action(self, action: TimelineRetentionAction) -> TimelineRetentionAction:
        self.db.add(action)
        self.db.flush()
        return action

    def add_tombstone(self, tombstone: TimelineTombstone) -> TimelineTombstone:
        self.db.add(tombstone)
        self.db.flush()
        return tombstone

    def list_retention_actions(
        self,
        *,
        search: str | None = None,
        action: str | None = None,
        mode: str | None = None,
        actor_id: str | None = None,
        entity_type: str | None = None,
        status_filter: str | None = None,
        page: int = 1,
        page_size: int = 25,
    ) -> tuple[list[TimelineRetentionAction], int]:
        conditions = []
        if action:
            conditions.append(TimelineRetentionAction.action == action)
        if mode:
            conditions.append(TimelineRetentionAction.mode == mode)
        if actor_id:
            conditions.append(TimelineRetentionAction.actor_id == actor_id)
        if entity_type:
            conditions.append(TimelineRetentionAction.entity_type == entity_type)
        if status_filter:
            conditions.append(TimelineRetentionAction.status == status_filter)
        if search and search.strip():
            term = f"%{search.strip()}%"
            conditions.append(or_(TimelineRetentionAction.reason.ilike(term), TimelineRetentionAction.actor_name.ilike(term)))
        total = self.db.scalar(select(func.count(TimelineRetentionAction.id)).where(*conditions)) or 0
        items = list(
            self.db.scalars(
                select(TimelineRetentionAction)
                .where(*conditions)
                .order_by(TimelineRetentionAction.created_at.desc())
                .offset((page - 1) * page_size)
                .limit(page_size)
            )
        )
        return items, total

    def due_retention_policies(self, now: datetime) -> list[TimelineRetentionPolicy]:
        return list(
            self.db.scalars(
                select(TimelineRetentionPolicy).where(
                    TimelineRetentionPolicy.is_active.is_(True),
                    TimelineRetentionPolicy.schedule_enabled.is_(True),
                    or_(TimelineRetentionPolicy.next_run_at.is_(None), TimelineRetentionPolicy.next_run_at <= now),
                )
            )
        )

    def add_handover_summary(self, summary: HandoverSummary) -> HandoverSummary:
        self.db.add(summary)
        self.db.flush()
        return summary

    def get_handover_summary(self, summary_id: str) -> HandoverSummary | None:
        return self.db.get(HandoverSummary, summary_id)

    def list_handover_summaries(
        self,
        account_id: str,
        page: int = 1,
        page_size: int = 25,
        *,
        search: str | None = None,
        generated_by_id: str | None = None,
        ownership_change_id: str | None = None,
        date_from: datetime | None = None,
        date_to: datetime | None = None,
    ) -> tuple[list[HandoverSummary], int]:
        conditions = [HandoverSummary.account_id == account_id]
        if generated_by_id:
            conditions.append(HandoverSummary.generated_by_id == generated_by_id)
        if ownership_change_id:
            conditions.append(HandoverSummary.ownership_change_id == ownership_change_id)
        if date_from:
            conditions.append(HandoverSummary.created_at >= date_from)
        if date_to:
            conditions.append(HandoverSummary.created_at <= date_to)
        if search and search.strip():
            term = f"%{search.strip()}%"
            conditions.append(
                or_(
                    HandoverSummary.generated_by_name.ilike(term),
                    cast(HandoverSummary.selected_sections, String).ilike(term),
                    cast(HandoverSummary.citations_json, String).ilike(term),
                    cast(HandoverSummary.content_json, String).ilike(term),
                )
            )
        total = self.db.scalar(select(func.count(HandoverSummary.id)).where(*conditions)) or 0
        items = list(
            self.db.scalars(
                select(HandoverSummary)
                .where(*conditions)
                .order_by(HandoverSummary.created_at.desc())
                .offset((page - 1) * page_size)
                .limit(page_size)
            )
        )
        return items, total

    def add_handover_share(self, share: HandoverShare) -> HandoverShare:
        self.db.add(share)
        self.db.flush()
        return share

    def add_ai_search_audit(self, audit: TimelineAiSearchAudit) -> TimelineAiSearchAudit:
        self.db.add(audit)
        self.db.flush()
        return audit

    @staticmethod
    def _timeline_conditions(
        *,
        account_id: str,
        search: str | None,
        event_types: list[str] | None,
        modules: list[str] | None,
        owner_id: str | None,
        source_record_type: str | None,
        source_record_id: str | None,
        status_filter: str | None,
        sensitivity_level: str | None,
        date_from: datetime | None,
        date_to: datetime | None,
        include_sensitive: bool,
        include_restricted: bool,
    ) -> list:
        conditions = [TimelineEntry.account_id == account_id, TimelineEntry.status != "deleted"]
        if status_filter:
            conditions.append(TimelineEntry.status == status_filter)
        if not include_restricted:
            conditions.append(TimelineEntry.status != "restricted")
        if not include_sensitive:
            conditions.append(TimelineEntry.is_sensitive.is_(False))
        if sensitivity_level:
            conditions.append(TimelineEntry.sensitivity_level == sensitivity_level)
        if event_types:
            conditions.append(TimelineEntry.event_type.in_(event_types))
        if modules:
            conditions.append(TimelineEntry.module.in_(modules))
        if owner_id:
            conditions.append(TimelineEntry.performed_by == owner_id)
        if source_record_type:
            conditions.append(TimelineEntry.source_record_type == source_record_type)
        if source_record_id:
            conditions.append(TimelineEntry.source_record_id == source_record_id)
        if date_from:
            conditions.append(TimelineEntry.event_at >= date_from)
        if date_to:
            conditions.append(TimelineEntry.event_at <= date_to)
        if search and search.strip():
            term = f"%{search.strip()}%"
            conditions.append(
                or_(
                    TimelineEntry.title.ilike(term),
                    TimelineEntry.description.ilike(term),
                    TimelineEntry.performed_by_name.ilike(term),
                    TimelineEntry.event_type.ilike(term),
                    TimelineEntry.module.ilike(term),
                    cast(TimelineEntry.tags, String).ilike(term),
                    cast(TimelineEntry.mentions, String).ilike(term),
                )
            )
        return conditions

    def commit(self) -> None:
        self.db.commit()


def policy_duration(policy: TimelineRetentionPolicy):
    from datetime import timedelta

    return timedelta(days=policy.duration_days)

File: frontend/src/services/timeline.ts
import { apiRequest } from '@/services/api'
import { TimelineComment, TimelineEntry, TimelineEventType, TimelineEventTypeConfig, TimelineModule, SensitivityLevel } from '@/types/timeline'

export interface Page<T> {
  items: T[]
  total: number
  page: number
  page_size: number
  pages: number
}

interface ApiTimelineEvent {
  id: string
  account_id: string
  engagement_id?: string | null
  event_type: string
  module?: string | null
  title: string
  description: string
  previous_value?: Record<string, unknown> | null
  new_value?: Record<string, unknown> | null
  before_value?: Record<string, unknown> | null
  after_value?: Record<string, unknown> | null
  actor_id: string
  actor_name: string
  performed_by?: string | null
  performed_by_name?: string | null
  source_module: string
  source_record_id?: string | null
  source_record_type?: string | null
  source_record_route?: string | null
  metadata?: Record<string, unknown> | null
  event_at?: string | null
  timestamp?: string | null
  is_sensitive: boolean
  sensitivity_level?: string | null
  tags: string[]
  mentions: string[]
  attachments: { name?: string; url?: string }[]
  status: string
  is_system_generated: boolean
  is_immutable: boolean
  created_at: string
  updated_at?: string | null
}

interface ApiTimelineComment {
  id: string
  timeline_entry_id: string
  author_id: string
  author_name: string
  body: string
  mentions: string[]
  created_at: string
  updated_at: string
}

interface ApiEventType {
  id: string
  slug: string
  name: string
  category: string
  module: string
  color_token: string
  display_order: number
  default_visibility: 'public' | 'restricted'
  retention_policy_id?: string | null
  is_active: boolean
  is_critical: boolean
  created_at: string
  updated_at: string
}

export interface TimelineFiltersApi {
  search?: string
  event_type?: string
  module?: string
  owner_id?: string
  date_from?: string
  date_to?: string
  show_sensitive?: boolean
  sort?: string
  direction?: 'asc' | 'desc'
  page?: number
  page_size?: number
}

export interface TimelineNotePayload {
  event_type: string
  title?: string
  description: string
  event_at?: string
  owner_id?: string
  mentions: string[]
  attachments: { name?: string; url?: string }[]
  is_sensitive?: boolean
  sensitivity_level?: string
  tags?: string[]
}

export interface TimelineAiSearchPayload {
  query: string
  scopes?: string[]
  document_search?: boolean
  date_from?: string
  date_to?: string
  limit?: number
}

export interface TimelineAiSearchResponse {
  query: string
  interpreted_intent: string
  mode: string
  confidence: 'high' | 'medium' | 'low'
  answer: string
  disclaimer: string
  results: {
    id: string
    title: string
    excerpt: string
    event_type: string
    source_module: string
    source_route?: string | null
    event_at: string
    relevance: number
    mode: string
  }[]
  document_results: {
    id: string
    source_label: string
    excerpt: string
    source_route?: string | null
  }[]
  audit_id: string
  can_try_in_kam_ai: boolean
}

export interface HandoverSummary {
  id: string
  account_id: string
  generated_by_id: string
  generated_by_name: string
  selected_sections: string[]
  source_set: { type: string; id: string }[]
  redaction_summary: Record<string, unknown>
  citations: { id: string; label: string; excerpt: string; source_route?: string | null }[]
  content: Record<string, unknown>
  status: string
  export_metadata: Record<string, unknown>
  share_metadata: Record<string, unknown>
  created_at: string
  updated_at: string
}

export interface HandoverShare {
  summary_id: string
  share_id: string
  internal_share_url: string
  created_at: string
  expires_at?: string | null
}

export interface RetentionPolicy {
  id: string
  name: string
  action: 'archive' | 'restrict' | 'delete'
  duration_days: number
  reason_template: string
  schedule_enabled: boolean
  schedule_interval_hours: number
  last_run_at?: string | null
  next_run_at?: string | null
  is_active: boolean
}

export interface RetentionAction {
  id: string
  policy_id?: string | null
  action: string
  mode: string
  status: string
  matched_count: number
  affected_count: number
  reason?: string | null
  actor_name: string
  created_at: string
}

export interface RetentionRunResult {
  policy_id: string
  action: string
  mode: string
  matched_count: number
  affected_count: number
  sample_event_ids: string[]
}

export async function getAccountTimeline(token: string, accountId: string, params: TimelineFiltersApi) {
  const query = queryString(params)
  const page = await apiRequest<Page<ApiTimelineEvent>>(`/api/accounts/${accountId}/timeline${query ? `?${query}` : ''}`, { token })
  return { ...page, items: page.items.map(mapTimelineEvent) }
}

export async function createTimelineNote(token: string, accountId: string, payload: TimelineNotePayload) {
  return mapTimelineEvent(await apiRequest<ApiTimelineEvent>(`/api/accounts/${accountId}/timeline-notes`, { method: 'POST', token, body: JSON.stringify(payload) }))
}

export async function updateTimelineEvent(token: string, eventId: string, payload: Partial<TimelineNotePayload>) {
  return mapTimelineEvent(await apiRequest<ApiTimelineEvent>(`/api/timeline-events/${eventId}`, { method: 'PATCH', token, body: JSON.stringify(payload) }))
}

export async function deleteTimelineEvent(token: string, eventId: string) {
  return apiRequest<{ message: string }>(`/api/timeline-events/${eventId}`, { method: 'DELETE', token })
}

export async function getTimelineComments(token: string, eventId: string, page = 1) {
  const result = await apiRequest<Page<ApiTimelineComment>>(`/api/timeline-events/${eventId}/comments?page=${page}&page_size=25`, { token })
  return { ...result, items: result.items.map(mapTimelineComment) }
}

export async function createTimelineComment(token: string, eventId: string, body: string, mentions: string[]) {
  return mapTimelineComment(await apiRequest<ApiTimelineComment>(`/api/timeline-events/${eventId}/comments`, { method: 'POST', token, body: JSON.stringify({ body, mentions }) }))
}

export async function updateTimelineComment(token: string, eventId: string, commentId: string, body: string, mentions: string[]) {
  return mapTimelineComment(await apiRequest<ApiTimelineComment>(`/api/timeline-events/${eventId}/comments/${commentId}`, { method: 'PATCH', token, body: JSON.stringify({ body, mentions }) }))
}

export async function deleteTimelineComment(token: string, eventId: string, commentId: string) {
  return apiRequest<{ message: string }>(`/api/timeline-events/${eventId}/comments/${commentId}`, { method: 'DELETE', token })
}

export async function getTimelineEventTypes(token: string, activeState = 'all') {
  const page = await apiRequest<Page<ApiEventType>>(`/api/timeline-event-types?active_state=${activeState}&page=1&page_size=100`, { token })
  return { ...page, items: page.items.map(mapEventType) }
}

export async function getRetentionPolicies(token: string) {
  return apiRequest<Page<RetentionPolicy>>('/api/admin/retention-policies?page=1&page_size=100', { token })
}

export async function createRetentionPolicy(token: string, payload: Partial<RetentionPolicy> & { name: string; action: string; duration_days: number }) {
  return apiRequest<RetentionPolicy>('/api/admin/retention-policies', { method: 'POST', token, body: JSON.stringify(payload) })
}

export async function updateRetentionPolicy(token: string, policyId: string, payload: Partial<RetentionPolicy>) {
  return apiRequest<RetentionPolicy>(`/api/admin/retention-policies/${policyId}`, { method: 'PATCH', token, body: JSON.stringify(payload) })
}

export async function simulateRetentionPolicy(token: string, policyId: string) {
  return apiRequest<RetentionRunResult>(`/api/admin/retention-policies/${policyId}/simulate`, { method: 'POST', token, body: JSON.stringify({ limit: 500 }) })
}

export async function runRetentionPolicy(token: string, policyId: string) {
  return apiRequest<RetentionRunResult>(`/api/admin/retention-policies/${policyId}/run`, { method: 'POST', token, body: JSON.stringify({ limit: 500 }) })
}

export async function getRetentionActions(token: string, params: { search?: string; action?: string; mode?: string; entity_type?: string; status?: string; page?: number; page_size?: number } = {}) {
  const query = queryString({ page: 1, page_size: 25, ...params })
  return apiRequest<Page<RetentionAction>>(`/api/admin/retention-actions?${query}`, { token })
}

export async function generateHandoverSummary(token: string, accountId: string, selectedSections: string[] = []) {
  return apiRequest<HandoverSummary>(`/api/accounts/${accountId}/handover-summary`, { method: 'POST', token, body: JSON.stringify({ selected_sections: selectedSections }) })
}

export async function getHandoverSummaries(token: string, accountId: string, params: { search?: string; generated_by_id?: string; ownership_change_id?: string; page?: number; page_size?: number } = {}) {
  const query = queryString({ page: 1, page_size: 25, ...params })
  return apiRequest<Page<HandoverSummary>>(`/api/accounts/${accountId}/handover-summaries?${query}`, { token })
}

export async function shareHandoverSummary(token: string, summaryId: string) {
  return apiRequest<HandoverShare>(`/api/handover-summaries/${summaryId}/share`, { method: 'POST', token })
}

export async function exportHandoverPdf(token: string, summaryId: string) {
  const base = import.meta.env.VITE_API_BASE_URL ?? 'http://127.0.0.1:8001'
  const response = await fetch(`${base}/api/handover-summaries/${summaryId}/export`, {
    method: 'POST',
    headers: { Authorization: `Bearer ${token}` },
  })
  if (!response.ok) throw new Error('Handover PDF could not be exported')
  return response.blob()
}

export async function runTimelineAiSearch(token: string, accountId: string, payload: TimelineAiSearchPayload) {
  return apiRequest<TimelineAiSearchResponse>(`/api/accounts/${accountId}/timeline/ai-search`, { method: 'POST', token, body: JSON.stringify(payload) })
}

export function handoverPdfUrl(summaryId: string) {
  const base = import.meta.env.VITE_API_BASE_URL ?? 'http://127.0.0.1:8001'
  return `${base}/api/handover-summaries/${summaryId}/export`
}

export function mapTimelineEvent(event: ApiTimelineEvent): TimelineEntry {
  return {
    id: event.id,
    accountId: event.account_id,
    eventType: toTimelineEventType(event.event_type),
    module: toTimelineModule(event.module ?? event.source_module),
    title: event.title,
    description: event.description,
    performedBy: event.performed_by ?? event.actor_id,
    performedByName: event.performed_by_name ?? event.actor_name,
    timestamp: event.event_at ?? event.timestamp ?? event.created_at,
    sourceRecordId: event.source_record_id ?? event.engagement_id ?? undefined,
    sourceRecordType: event.source_record_type ?? (event.engagement_id ? 'engagement' : undefined),
    sourceRecordRoute: event.source_record_route ?? undefined,
    beforeValue: event.before_value ?? event.previous_value ?? undefined,
    afterValue: event.after_value ?? event.new_value ?? undefined,
    isSensitive: event.is_sensitive,
    sensitivityLevel: toSensitivity(event.sensitivity_level),
    tags: event.tags ?? [],
    mentions: event.mentions ?? [],
    attachments: event.attachments?.map(item => ({ name: item.name ?? 'Attachment', url: item.url ?? '' })).filter(item => item.url),
    metadata: event.metadata ?? undefined,
    isSystemGenerated: event.is_system_generated,
    isImmutable: event.is_immutable,
    retentionPolicy: event.status === 'archived' ? 'archive' : undefined,
  }
}

function mapTimelineComment(comment: ApiTimelineComment): TimelineComment {
  return {
    id: comment.id,
    entryId: comment.timeline_entry_id,
    authorId: comment.author_id,
    authorName: comment.author_name,
    content: comment.body,
    timestamp: comment.created_at,
    mentions: comment.mentions ?? [],
  }
}

function mapEventType(item: ApiEventType): TimelineEventTypeConfig {
  return {
    id: item.id,
    name: item.name,
    eventType: toTimelineEventType(item.slug),
    module: toTimelineModule(item.module),
    colorToken: item.color_token,
    active: item.is_active,
    defaultVisibility: item.default_visibility,
    createdDate: item.created_at,
    retentionPolicy: 'keep',
  }
}

function toTimelineEventType(value: string): TimelineEventType {
  return value as TimelineEventType
}

function toTimelineModule(value?: string | null): TimelineModule {
  const module = value || 'manual'
  return module as TimelineModule
}

function toSensitivity(value?: string | null): SensitivityLevel | undefined {
  if (value === 'commercial' || value === 'executive' || value === 'legal' || value === 'escalation') return value
  return undefined
}

function queryString(params: object) {
  const query = new URLSearchParams()
  Object.entries(params).forEach(([key, value]) => {
    if (value === undefined || value === null || value === '') return
    query.set(key, String(value))
  })
  return query.toString()
}
